cli/cmd/metrics_diagnostics_util.go

```diff
@@ -0,0 +1,144 @@
+package cmd
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"sort"
+	"sync/atomic"
+	"time"
+
+	"github.com/linkerd/linkerd2/pkg/k8s"
+	corev1 "k8s.io/api/core/v1"
+)
+
+// shared between metrics and diagnostics command
+type metricsResult struct {
+	pod       string
+	container string
+	metrics   []byte
+	err       error
+}
+type byResult []metricsResult
+
+func (s byResult) Len() int {
+	return len(s)
+}
+func (s byResult) Swap(i, j int) {
+	s[i], s[j] = s[j], s[i]
+}
+func (s byResult) Less(i, j int) bool {
+	return s[i].pod < s[j].pod || ((s[i].pod == s[j].pod) && s[i].container < s[j].container)
+}
+
+// getResponse makes a http Get request to the passed url and returns the response/error
+func getResponse(url string) ([]byte, error) {
+	resp, err := http.Get(url)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+	return ioutil.ReadAll(resp.Body)
+}
+
+// getContainerMetrics returns the metrics exposed by a container on the passed in portName
+func getContainerMetrics(
+	k8sAPI *k8s.KubernetesAPI,
+	pod corev1.Pod,
+	container corev1.Container,
+	emitLogs bool,
+	portName string,
+) ([]byte, error) {
+	portForward, err := k8s.NewContainerMetricsForward(k8sAPI, pod, container, emitLogs, portName)
+	if err != nil {
+		return nil, err
+	}
+
+	defer portForward.Stop()
+	if err = portForward.Init(); err != nil {
+		fmt.Fprintf(os.Stderr, "Error running port-forward: %s", err)
+		return nil, err
+	}
+
+	metricsURL := portForward.URLFor("/metrics")
+	return getResponse(metricsURL)
+}
+
+// getAllContainersWithPort returns all the containers within
+// a pod which exposes metrics at a port with name portName
+func getAllContainersWithPort(
+	pod corev1.Pod,
+	portName string,
+) ([]corev1.Container, error) {
+	if pod.Status.Phase != corev1.PodRunning {
+		return nil, fmt.Errorf("pod not running: %s", pod.GetName())
+	}
+	var containers []corev1.Container
+
+	for _, c := range pod.Spec.Containers {
+		for _, p := range c.Ports {
+			if p.Name == portName {
+				containers = append(containers, c)
+			}
+		}
+	}
+	return containers, nil
+}
+
+// getMetrics returns the metrics exposed by all the containers of the passed in list of pods
+// which exposes their metrics at portName
+func getMetrics(
+	k8sAPI *k8s.KubernetesAPI,
+	pods []corev1.Pod,
+	portName string,
+	waitingTime time.Duration,
+	emitLogs bool,
+) []metricsResult {
+	var results []metricsResult
+
+	resultChan := make(chan metricsResult)
+	var activeRoutines int32
+	for _, pod := range pods {
+		containers, err := getAllContainersWithPort(pod, portName)
+		if err != nil {
+			resultChan <- metricsResult{
```

Comment:
Unfortunately, this is going to cause the sender to block, because the `resultChan` hasn't been set up to receive data yet. You can test it by changing the conditional to `if err == nil`. I think the easiest solution here is to turn `resultChan` into a buffered channel: 

```diff
diff --git a/cli/cmd/metrics_diagnostics_util.go b/cli/cmd/metrics_diagnostics_util.go
index 49dd86f0..470863a9 100644
--- a/cli/cmd/metrics_diagnostics_util.go
+++ b/cli/cmd/metrics_diagnostics_util.go
@@ -97,7 +97,7 @@ func getMetrics(
 ) []metricsResult {
 	var results []metricsResult
 
-	resultChan := make(chan metricsResult)
+	resultChan := make(chan metricsResult, len(pods))
 	var activeRoutines int32
 	for _, pod := range pods {
 		containers, err := getAllContainersWithPort(pod, portName)
```

Alternately, just spawn a goroutine to handle each pod, and don't bother with the goroutine at line 114 below.

Reply:
Yeah changing it to a buffered channel was helpful.

Although after changing it to a buffered channel I faced another problem because of line 136 where we check if the `activeRoutines` is zero or not. After I changed the conditional to `if err == nil`, I was only getting 1 entry in the `results` slice since the `for` loop of line 129 was exiting after receiving one entry from the `resultsChan` since `activeRoutines` was still zero. 

In order to deal with it, I simply removed the `activeRoutines` variable and declared a `expectedLen` variable indicating the expected length of the results.

Reply:
> After I changed the conditional to if err == nil

That is meant to reproduce the blocking problem. We still want to keep it as `if err != nil`.

Reply:
I think you misunderstood. I'll explain the problem in detail now from the start. I do understand that we still want to keep `if err != nil`.

Consider this snippet

```golang
resultChan := make(chan metricsResult, len(pods))
var activeRoutines int32
for _, pod := range pods {
	containers, err := getAllContainersWithPort(pod, portName)
	if err != nil {
		resultChan <- metricsResult{
			pod: pod.GetName(),
			err: err,
		}
		continue
	}

	for i := range containers {
		atomic.AddInt32(&activeRoutines, 1)
		go func(p corev1.Pod, c corev1.Container) {
			defer atomic.AddInt32(&activeRoutines, -1)
			// get metrics and send to channel
		}(pod, containers[i])
	}
}

for {
	select {
	case result := <-resultChan:
		results = append(results, result)
	case <-time.After(waitingTime):
		break // timed out
	}
	if atomic.LoadInt32(&activeRoutines) == 0 {
		break
	}
}
```

Imagine if there was an error in getting the containers for all of the pods i.e. the `getAllContainersWithPort` method returns an error for each pod of the `pods` slice. This would mean that the `activeRoutines` variable stays set to 0. Now consider the `for` loop where we check if the `activeRoutines` variable is 0 or not.
In this particular case, the `if` condition will be satisfied and the loop will break just after getting one entry from the `resultsChan`.

Reply:
> Alternately, just spawn a goroutine to handle each pod, and don't bother with the goroutine at line 114 below.

This works well and doesn't have the above problem. Hence, I have gone for this option

Reply:
Glad the second approach works. With the first approach, there are ways to determine if a buffered channel still has any unread elements.